---
title:  "Fst sliding analysis  - divergence of N-MT genes"
author: "Natalia Zajac"
output:
  html_document: 
  highlight: pygments
theme: sand
code_folding: hide
toc: yes
toc_depth: 4
editor_options: 
  chunk_output_type: inline
---

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

```{r setup, include=FALSE}
library(poolfstat)
library(pheatmap)
library(dplyr)
library(cowplot)
library(patchwork)
library(readr)
library(stringr)
library(tidyverse)
library(textshaping)
library(ggVennDiagram)
library(ComplexHeatmap)
library(data.table)
library(ezRun)
library(scales)
library(GSEABase)
library(tidyverse)
library(rstatix)
library(ggpubr)
library(ggrepel)

Fst_grendalf = read_delim("../FST/Popoolation2/Sliding_Fst_per_Lake/Fst_sliding.Grendalf.Pi.FET.C2.txt")

df = read_delim("../test.sync") # just for colnames
poolnames = str_remove(sapply(str_split(colnames(df)[c(-1:-3)], "\\."), .subset, 2), "B-")
metadata = data.frame(pool = poolnames , lake = sapply(str_split(poolnames, "_"), .subset, 1)) %>% column_to_rownames("pool") %>% mutate(lake = if_else(lake == "Alex", "Alexandrina", lake))

colors = list(lake = setNames(c("orange", "cornflowerblue","gold", "purple", "darkgreen"), unique(metadata$lake)))
geneAnno = read_delim("../Atriophallophorus_winterbourni/EBI/GCA_013407085.1/Annotation/Genes/genes_annotation_byGene.txt")

genes = data.table(geneAnno[,c(6,8,9)])
colnames(genes) = c("Chr", "start", "end")
setkey(genes, Chr, start, end)

snps = Fst_grendalf[,c(1,2,3)] %>% unique() 
snps = data.table(snps)
setkey(snps, chrom, start, end)
obj = foverlaps(snps, genes, by.x = c("chrom", "start", "end"), 
                by.y = c("Chr", "start", "end"), type = "within", nomatch = 0)
obj = merge(data.frame(obj), geneAnno[,c(3,6,8,9)], by.x = c("chrom", "start", "end"), by.y = c("seqid", "start", "end"), all = TRUE)
Fst_grendalf = merge(Fst_grendalf, obj %>% select(-2,-3) , by.x = c("chrom", "start", "end"), by.y = c("chrom", "i.start", "i.end"), all.x = TRUE) %>% replace(is.na(.), "")
Fst_grendalf$window = paste0(Fst_grendalf$chrom, "-", Fst_grendalf$start, "-", Fst_grendalf$end)


library(GO.db)
library(enrichR)
library(clusterProfiler)
term2gene = NULL
term2gene[['BP']] = geneAnno[,c(3,12)] %>% filter(!is.na(`GO BP`)) %>%
    separate_rows(`GO BP`, sep = "; ") %>%
    dplyr::rename(Term = `GO BP`, Gene = gene_name) %>% dplyr::select(c(Term,Gene))
term2gene[['MF']] = geneAnno[,c(3,13)] %>% filter(!is.na(`GO MF`)) %>%
    separate_rows(`GO MF`, sep = "; ") %>%
    dplyr::rename(Term = `GO MF`, Gene = gene_name) %>% dplyr::select(c(Term,Gene))
term2gene[['CC']] = geneAnno[,c(3,14)] %>% filter(!is.na(`GO CC`)) %>%
    separate_rows(`GO CC`, sep = "; ") %>%
    dplyr::rename(Term = `GO CC`, Gene = gene_name) %>% dplyr::select(c(Term,Gene))

term2name= NULL
term2name[["BP"]] = data.frame(Term = term2gene$BP$Term, Name = sapply(term2gene$BP$Term, function(x){Term(x)[[1]]}))
term2name[["MF"]] = data.frame(Term = term2gene$MF$Term, Name = sapply(term2gene$MF$Term, function(x){Term(x)[[1]]}))
term2name[["CC"]] = data.frame(Term = term2gene$CC$Term, Name = sapply(term2gene$CC$Term, function(x){Term(x)[[1]]}))

IC = read.delim("../Atriophallophorus_winterbourni/EBI/GCA_013407085.1/GO_files/Atrio_GO_IC", header = F, sep = " ")

#Add KO annotation
kegg = read.delim("../Gene2Kegg_Awinterbourni.txt",header = F, sep = "\t")
kegg = kegg[-1,c(1,2,3)]
term2gene$Kegg = kegg[kegg$V2 != "",c(2,1)]
colnames(term2gene$Kegg) = c("KO","Gene")
term2gene$Kegg$Gene = str_remove(term2gene$Kegg$Gene, "-mRNA-1")
term2name$Kegg = kegg[kegg$V2 != "",c(2,3)]
colnames(term2name$Kegg) = c("KO", "Name")

for_plots = NULL
for (i in unique(Fst_grendalf$label)[c(2:5,7:10,11,15)]){
  tb = Fst_grendalf[Fst_grendalf$label == i,]
  x = list("Grendalf" = tb[tb$MeanGrendalfFst_4sd == "outlier",]$window, 
           "C2" = tb[tb$C2std_Sig_SNPs > 0,]$window, 
           "FetMean" = tb[tb$FET_SigMean_SNPs > 0,]$window, 
           "FetMedian" = tb[tb$FET_SigMedian_SNPs > 0,]$window)
  tb$color = if_else(tb$window %in% Reduce(intersect, x), "final_outlier", "not_outlier")
  for_plots = rbind(for_plots, tb)
}

for_plots = for_plots %>% mutate(label2 = case_when(label %in% c("Alexandrina_Paringa", "Alexandrina_Mapourika", "Middleton_Paringa", "Mapourika_Middleton") ~ "West vs South Canterbury", 
                                        label %in% c("Alexandrina_Selfe", "Middleton_Selfe") ~ "South Canterbury vs Selfe", 
                                        label %in% c("Mapourika_Selfe", "Paringa_Selfe") ~ "West vs Selfe", 
                                        label == "Alexandrina_Middleton" ~ "East vs East",  label == "Mapourika_Paringa" ~ "West vs West"))

#Annotate Gos with GOSSlims
library(GSEABase)
library(tidyverse)
library(rstatix)

og_slimdf = NULL
for (i in c("BP", "MF", "CC")){
  myIds <- term2gene[[i]]$Term
  myCollection <- GOCollection(myIds)
  fl <- "../Atriophallophorus_winterbourni/EBI/GCA_013407085.1/GO_files/goslim.obo" #go_slim_generic
  slim <- getOBOCollection(fl)
  slimdf <- goSlim(myCollection, slim, ontology = i)
  if (i == "BP"){
    GO.db::GOBPOFFSPRING
      mappedIds <-
        function(df, collection, OFFSPRING)
        {
        map <- as.list(OFFSPRING[rownames(df)])
        mapped <- lapply(map, intersect, ids(collection))
        df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L))
        df
        }
      slimdf <- mappedIds(slimdf, myCollection, GOBPOFFSPRING)
  } else if (i == "MF"){
     GO.db::GOMFOFFSPRING
      mappedIds <-
        function(df, collection, OFFSPRING)
          {
          map <- as.list(OFFSPRING[rownames(df)])
          mapped <- lapply(map, intersect, ids(collection))
          df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L))
          df
          }
      slimdf <- mappedIds(slimdf, myCollection, GOMFOFFSPRING)
  } else {
    GO.db::GOCCOFFSPRING
      mappedIds <-
        function(df, collection, OFFSPRING)
          {
          map <- as.list(OFFSPRING[rownames(df)])
          mapped <- lapply(map, intersect, ids(collection))
          df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L))
          df
          }
      slimdf <- mappedIds(slimdf, myCollection, GOCCOFFSPRING)
  }
  slimdf = slimdf %>% rownames_to_column("GOslim") %>% separate_longer_delim(go_terms, delim = ";")
  slimdf = merge(merge(slimdf, term2gene[[i]], by.x = "go_terms", by.y = "Term"), term2name[[i]], by.x = "go_terms", by.y = "Term")
  slimdf$GO = i
  og_slimdf = rbind(og_slimdf,slimdf)
}

Cb64k <- c("#000000", "#FFFF00", "#1CE6FF", "#FF34FF", "#FF4A46", "#008941", "#006FA6", "#A30059",
           "#FFDBE5", "#7A4900", "#0000A6", "#63FFAC", "#B79762", "#004D43", "#8FB0FF", "#997D87",
           "#5A0007", "#809693", "#FEFFE6", "#1B4400", "#4FC601", "#3B5DFF", "#4A3B53", "#FF2F80",
           "#61615A", "#BA0900", "#6B7900", "#00C2A0", "#FFAA92", "#FF90C9", "#B903AA", "#D16100",
           "#DDEFFF", "#000035", "#7B4F4B", "#A1C299", "#300018", "#0AA6D8", "#013349", "#00846F",
           "#372101", "#FFB500", "#C2FFED", "#A079BF", "#CC0744", "#C0B9B2", "#C2FF99", "#001E09",
           "#00489C", "#6F0062", "#0CBD66", "#EEC3FF", "#456D75", "#B77B68", "#7A87A1", "#788D66",
           "#885578", "#FAD09F", "#FF8A9A", "#D157A0", "#BEC459", "#456648", "#0086ED", "#886F4C",
           "#34362D", "#B4A8BD", "#00A6AA", "#452C2C", "#636375", "#A3C8C9", "#FF913F", "#938A81",
           "#575329", "#00FECF", "#B05B6F", "#8CD0FF", "#3B9700", "#04F757", "#C8A1A1", "#1E6E00",
           "#7900D7", "#A77500", "#6367A9", "#A05837", "#6B002C", "#772600", "#D790FF", "#9B9700",
           "#549E79", "#FFF69F", "#201625", "#72418F", "#BC23FF", "#99ADC0", "#3A2465", "#922329",
           "#5B4534", "#FDE8DC", "#404E55", "#0089A3", "#CB7E98", "#A4E804", "#324E72", "#6A3A4C")
```


# {.tabset}

This notebook is for Fst sliding window analysis without Paringa Pool4. This analysis will be top down - classifying the genes into clusters by GO annotation and checking the average Fst/ proportion of outliers per functional category. The idea came from Fst_sliding_window_part3.

One thing you have to remember - you are trying to see from the genomic data if the more likely hypothesis is mit-nuclear incompatibility or assymetric sexual reproduction (females from West, males from East) - which in hermaphordites can be caused by sexual conflict, sex biased offspring production, sequential hermaphroditism, see literature.


## Papers

Sex specific orthologs from all trematodes and mitochondrial genes

```{r orths, include=FALSE}
sex_specific_orthologs = read_delim("../Sex_specific_genes/Common_results/Sex_specific_genes_final_dataframe.txt")

temp = merge(for_plots, sex_specific_orthologs, by.x = "gene_name", by.y = "Atrio", all.x = TRUE) %>% replace(is.na(.), "") %>% mutate(mitochondrial = if_else(gene_name %in% mtcandidatesT, "Mitochondrial", ""))
temp = temp %>% mutate(Final_label = case_when(mitochondrial != "" & Label !=  "" ~ paste0(Label, ", ", mitochondrial), mitochondrial != "" & Label ==  "" ~ mitochondrial, mitochondrial == "" & Label !=  "" ~ Label)) %>% replace(is.na(.), "")
```

Lets look  at the mitochondrial genes in more detail

```{r, fig.height=10, fig.width=15}
#https://www.mdpi.com/2218-273X/12/3/427#B164-biomolecules-12-00427

mtcandidatesT = list("KEGG"= unique(term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("mitoch|acetyl-CoA|fumarate|riboflavin|MECR.*mitoch|AMPK|target of rapamycin|SIRT|SDH|NAD|BCS1|FDX2|iron-sulfur|reticulon|SLC25|mitoguardin|MAPK|MAP2K|MORG1|JNK", term2name$Kegg$Name),]$KO,]$Gene), 
  "GOslim" = c(unique(og_slimdf[grepl("mitochondri",og_slimdf$Term),]$Gene), 
  unique(og_slimdf[grepl("NADH dehydrogenase",og_slimdf$Name),]$Gene)))

mt = og_slimdf[grepl("mitochondri",og_slimdf$Term),]
mt2 = og_slimdf[grepl("NADH dehydrogenase",og_slimdf$Name),]
mt = rbind(mt,mt2)

list_of_mt_categories = list(
  "mitochondrial fusion" = c(unique(mt[grepl("fusion", mt$Name),]$Gene),term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("mitofusin", term2name$Kegg$Name),]$KO,]$Gene),
  "mitochondrial fission" = c(unique(mt[grepl("fission", mt$Name),]$Gene),term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("fission", term2name$Kegg$Name),]$KO,]$Gene),
  "ATP synthase" = c(unique(mt[grepl("ATP synthase", mt$Name),]$Gene), term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("ATP synthase", term2name$Kegg$Name),]$KO,]$Gene),
  "ERMES complex" = c(unique(mt[grepl("ER", mt$Name),]$Gene)),
  "mitochondrial membrane organization and function" = c(unique(mt[grepl("TIM|membrane|cristae|PAM|MICOS|SAM|crista", mt$Name),]$Gene),term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("TIM.*membrane|Tim|mitochondrial inner membrane|mitochondrial import|ERV1|ATAD1|PMPCA", term2name$Kegg$Name),]$KO,]$Gene),
  "mitochondrial respiratory chain complex" =  unique(mt[grepl("respiratory", mt$Name),]$Gene),
  "mitochondrial transcription and translation" = c(unique(mt[grepl("transcription|translation|polymerase", mt$Name),]$Gene), term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("polymerase.*mitoch|polymerase.*calcium|MTERF|TFAM", term2name$Kegg$Name),]$KO,]$Gene),
  "cytochrome c assembly and regulation" = c(unique(mt[grepl("cytochrome", mt$Name),]$Gene),term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("COX.*mitoch|mitocho.*COX", term2name$Kegg$Name),]$KO,]$Gene),
  "Ca2+ homeostasis" = c(term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("calcium.*mitoch|mitocho.*calcium|LETM1|SMDT1", term2name$Kegg$Name),]$KO,]$Gene, unique(mt[grepl("uniplex complex", mt$Name),]$Gene)),
  "mitochondrial apoptosis" = c(term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("JNK|ENDOG", term2name$Kegg$Name),]$KO,]$Gene, unique(mt[grepl("apop|autophagy|mitophagy|cell death", mt$Name),]$Gene)),
  "metabolic sensors" = c(term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("AMPK|target of rapamycin|SIRT", term2name$Kegg$Name),]$KO,]$Gene),
  "mitochondrial tRNA synthesis and modification" = unique(mt[grepl("RNA|ribonuclease P complex", mt$Name),]$Gene),
  "mitochondrial DNA maintainance and regulation" = c(unique(mt[grepl("DNA|genome maintenance|gamma DNA|kinetoplast|nucleoid", mt$Name),]$Gene)),
  "TCA cycle" = c(term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("acetyl-CoA|fumarate|riboflavin|MECR.*mitoch", term2name$Kegg$Name),]$KO,]$Gene, unique(mt[grepl("ketoglutarate|pyruvate dehydrogenase|oxoglutarate dehydrogenase|tricarboxylic acid cycle", mt$Name),]$Gene)),
  "ETC" = c(term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("SDH|NAD|BCS1|FDX2|iron-sulfur|reticulon", term2name$Kegg$Name),]$KO,]$Gene,unique(mt[grepl("succinate|NAD|electron transfer flavoprotein|respirasome", mt$Name),]$Gene)),
  "fatty acid beta-oxidation" = c(unique(mt[grepl("fatty acid beta-oxidation", mt$Name),]$Gene)),
  "SLC25 Family" = c(term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("SLC25", term2name$Kegg$Name),]$KO,]$Gene),
  "MAPK proteins" = c(term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("MAPK|MAP2K|MORG1|JNK", term2name$Kegg$Name),]$KO,]$Gene),
  "mitochondrial transport" = c(unique(mt[grepl("distribution", mt$Name),]$Gene), term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("MTCH|RHOT1", term2name$Kegg$Name),]$KO,]$Gene),
  "mitochondrial lipid metabolism" = c(term2gene$Kegg[term2gene$Kegg$KO %in% term2name$Kegg[grepl("LACTB|mitoguardin", term2name$Kegg$Name),]$KO,]$Gene,unique(mt[grepl("prohibitin", mt$Name),]$Gene)),
  "protein synthesis and degredation"= unique(mt[grepl("Cdc48p-Npl4p-Vms1p AAA ATPase complex|i-AAA complex|mitochondrial processing peptidase complex|mitochondrial ribosome", mt$Name),]$Gene)
)

mtcandidatesT <- lapply(names(mtcandidatesT), function(name) {
    data <- data.frame("Gene" = mtcandidatesT[[name]])
    data$Annotation <- name
    return(data)
})
mtcandidatesT = bind_rows(mtcandidatesT)


list_of_mt_categories = lapply(names(list_of_mt_categories), function(name) {
    data <- data.frame("Gene" = list_of_mt_categories[[name]])
    data$Function <- name
    return(data)
})
list_of_mt_categories = bind_rows(list_of_mt_categories)


mtsummary = merge(mtcandidatesT, list_of_mt_categories, by = "Gene", all = TRUE) %>% replace(is.na(.), "Miscellaneous") 
gene_annotation_counts <- mtsummary %>%
    group_by(Gene) %>%
    dplyr::summarize(Annotation_Category = ifelse(n_distinct(Annotation) > 1, "Both", Annotation[1])) 
mtsummary <- mtsummary %>%
    left_join(gene_annotation_counts, by = "Gene") %>%
    distinct(Gene, Function, Annotation_Category)
mtsummary <- mtsummary %>%
    group_by(Function, Annotation_Category) %>%
    dplyr::summarize(Unique_Gene_Count = n_distinct(Gene), .groups = "drop")

mtsummary = mtsummary %>% mutate(Category = if_else(Function %in% c("ATP synthase","mitochondrial respiratory chain complex","SLC25 Family", "MAPK proteins", "metabolic sensors", "ERMES complex"), "Component", "Function"))

mtsummary$Function = factor(mtsummary$Function, levels = c("SLC25 Family", "mitochondrial respiratory chain complex", "MAPK proteins", "ATP synthase", "ERMES complex", "metabolic sensors","TCA cycle", "ETC", "fatty acid beta-oxidation", "mitochondrial fusion", "mitochondrial fission", "mitochondrial apoptosis", "mitochondrial transport", "mitochondrial lipid metabolism", "mitochondrial transcription and translation", "mitochondrial tRNA synthesis and modification", "mitochondrial membrane organization and function", "mitochondrial DNA maintainance and regulation", "cytochrome c assembly and regulation", "Ca2+ homeostasis", "protein synthesis and degredation", "Miscellaneous"))

total_counts_df <- mtsummary %>%
    group_by(Function, Category) %>%  
    dplyr::summarise(Total_Count = sum(Unique_Gene_Count), .groups = "drop")


#Load these tables from files if the above produces weird result

mtsummary = read.delim("mtsummary.txt")
total_counts_df <- mtsummary %>%
    group_by(Function, Category) %>%  
    dplyr::summarise(Total_Count = sum(Unique_Gene_Count), .groups = "drop")
list_of_mt_categories = read.delim("list_of_mt_categories.txt")
og_slimdf = read.delim("ogslimdf.txt")

```



```{r, fig.height=10, fig.width=15}
pdf("../Figures/GeneCount_MitochondrialCategories.pdf", height = 7, width = 11)
ggplot(mtsummary, aes(Unique_Gene_Count, str_wrap(Function,30), fill = Annotation_Category)) + 
    geom_col() + 
    facet_grid(Category ~ ., scales = "free", space = "free") + 
    geom_text(data = total_counts_df, aes(x = Inf, y = str_wrap(Function,30), label = Total_Count), inherit.aes = FALSE, hjust = 1.1) + 
    scale_x_continuous(expand = expansion(mult = c(0, 0.2))) + 
    labs(y = "", x = "Count", fill = "Annotation") + 
    theme(strip.text = element_text(size = 15), legend.position = "top", legend.title = element_text(size = 11), legend.text = element_text(size = 11), axis.text.y = element_text(size = 11))

function_levels = mtsummary[,c(1,4)] %>% unique() %>% arrange(Category) %>% dplyr::select(Function) 
function_levels = function_levels$Function
mtsummary$Function = factor(mtsummary$Function, levels = function_levels)
funcols = setNames(brewPalette(n=22),levels(mtsummary$Function))

ggplot(mtsummary, aes(Function, Unique_Gene_Count, fill = factor(Function, levels = function_levels))) + geom_bar(stat = "identity", position = "dodge") + theme_bw() +
    theme(legend.text = element_text(size = 9), axis.text.x = element_blank(), axis.title.x = element_blank()) + scale_fill_manual(values = funcols) + facet_grid(~Category, scales = "free", space = "free") + guides(fill=guide_legend(ncol=1)) + labs(fill = "N-MT gene category", y = "Unique gene count") + theme(text = element_text(size = 15), legend.text = element_text(size = 10)) + 
    geom_text(data = total_counts_df, aes(x = factor(Function, levels = function_levels), y = Inf, label = Total_Count), inherit.aes = FALSE, vjust = 1.5)
dev.off()
```


```{r, fig.height=10, fig.width=15}

xx = merge(for_plots[,c(1:5,10,12,18,19,20,21)], list_of_mt_categories, by.x = "gene_name", by.y = "Gene", all.x = TRUE) %>% mutate(Function = if_else(is.na(Function) == TRUE, "Rest", Function))
xx = xx %>% mutate(Function2 = if_else(Function == "Rest", "Rest", "Mitochondrial"))
xx = xx %>% mutate(label2 = case_when(label %in% c("Alexandrina_Paringa", "Alexandrina_Mapourika", "Middleton_Paringa", "Mapourika_Middleton") ~ "West vs South Canterbury", label %in% c("Alexandrina_Selfe", "Middleton_Selfe") ~ "South Canterbury vs Selfe",  label %in% c("Mapourika_Selfe", "Paringa_Selfe") ~ "West vs Selfe",  label == "Alexandrina_Middleton" ~ "within South Canterbury",  label == "Mapourika_Paringa" ~ "within West"))
xx$label2 = factor(xx$label2, levels = c("within West", "West vs Selfe", "West vs South Canterbury", "South Canterbury vs Selfe", "within South Canterbury"))

distance = "../tLCP_1400.txt"
distance = read.delim(distance, sep = " ")
distance = distance[rownames(distance) %in% unique(metadata$lake),colnames(distance) %in% unique(metadata$lake)] %>% rownames_to_column() %>% reshape2::melt()
distance = distance[!duplicated(distance),]
distance$label = paste0(distance$rowname, "_", distance$variable)
distance$distance = distance$value
distance = distance[,c(5,4)]
xx = merge(xx, distance, by = "label", all.x = TRUE)
xxx = xx %>% arrange(chrom,start, end) 
```

Heatmap of Fst per function divided by average pairwise Fst (sort of normalized)
1 - average Fst of that function == average pairwise Fst
<1 - average Fst of that function < average pairwise Fst
>1 - average Fst of that function > average pairwise Fst

```{r, fig.height=6, fig.width=13}
pheatmap(merge(xx %>% group_by(gene_name, label2, Function) %>% summarise(Fst = mean(Fst)), for_plots %>% group_by(label2) %>% summarise(meanFst = mean(Fst)), by = "label2") %>% group_by(label2, Function) %>% summarise(meanFst = mean(Fst/meanFst)) %>% pivot_wider(values_from = meanFst, names_from = Function) %>% column_to_rownames("label2") %>% t())
```

Heatmap of the average proportion of outliers per category.

```{r, fig.height=6, fig.width=13}

pheatmap(xx %>% group_by(gene_name, label2, color, Function) %>% summarise(count = n_distinct(window)) %>% pivot_wider(values_from = count, names_from = color) %>% replace(is.na(.), 0) %>% mutate(sum = not_outlier + final_outlier, not_outlier = not_outlier/sum, final_outlier = final_outlier/sum) %>% group_by(label2, Function) %>% summarise(med = mean(final_outlier)) %>% pivot_wider(names_from = label2, values_from = med) %>% column_to_rownames("Function"))

```

Average proportion of outliers

```{r, fig.height=5, fig.width=6}

pdf("../Figures/ProportionOutliers_FST_Mitochondrial.pdf", width = 4, height = 8)
p1 = xx %>% mutate(Function2 = if_else(Function == "Rest", "Non-Mitochondrial", "Mitochondrial")) %>% group_by(color, label, label2, gene_name, Function2) %>% summarise(nb_windows = n_distinct(window)) %>% pivot_wider(names_from = color, values_from = nb_windows) %>% replace(is.na(.), 0) %>% mutate(prop_final_out = final_outlier/(final_outlier+not_outlier)) %>% group_by(label2, Function2) %>% summarise(prop_final_out = mean(prop_final_out)) %>% ggplot(aes(prop_final_out,str_wrap(label2,10), color = Function2)) + geom_point(size = 3) + labs(x = "Average proportion of outliers", y = "", color = "category") + theme(plot.margin = unit(c(5,5,5,5),"mm"), axis.text.y = element_text(size = 8), legend.text = element_text(size = 8), legend.position = "top") + scale_color_manual(values = c("black","darkgray"))

p2 = xx %>% mutate(Function2 = if_else(Function == "Rest", "Non-Mitochondrial", "Mitochondrial")) %>% group_by(label2, Function2) %>% summarise(Fst = mean(Fst)) %>% ggplot(aes(Fst, str_wrap(label2,10),color = Function2)) + geom_point(size = 3) + labs(y = "", color = "category") + xlab(expression(F[ST])) + theme(plot.margin = unit(c(5,5,5,5),"mm"), axis.text.y = element_text(size = 8), legend.position = "", legend.text = element_text(size = 8)) + scale_color_manual(values = c("black","darkgray"))
p1/p2
dev.off()
```

Average proportion of outliers for per lake comparison

```{r, fig.height=5, fig.width=6}
xx %>% mutate(Function2 = if_else(Function == "Rest", "Rest", "Mitochondrial")) %>% group_by(color, label, label2, gene_name, Function2) %>% summarise(nb_windows = n_distinct(window)) %>% pivot_wider(names_from = color, values_from = nb_windows) %>% replace(is.na(.), 0) %>% mutate(prop_final_out = final_outlier/(final_outlier+not_outlier)) %>% group_by(label, Function2) %>% summarise(prop_final_out = mean(prop_final_out)) %>% ggplot(aes(prop_final_out,label, color = Function2)) + geom_point()+ labs(x = "Average proportion of outliers", y = "", color = "category") + theme(legend.position = "top")
```

Average Fst

```{r, fig.height=5, fig.width=6}
xx %>% mutate(Function2 = if_else(Function == "Rest", "Rest", "Mitochondrial")) %>% group_by(label2, Function2) %>% summarise(Fst = mean(Fst)) %>% ggplot(aes(Fst, label2,color = Function2)) + geom_point() + labs(x = "FST", y = "", color = "category") + theme(legend.position = "top")
```

Average Fst per lake comparison

```{r, fig.height=5, fig.width=6}
xx %>% mutate(Function2 = if_else(Function == "Rest", "Rest", "Mitochondrial")) %>% group_by(label, Function2) %>% summarise(Fst = mean(Fst)) %>% ggplot(aes(Fst, label,color = Function2)) + geom_point() + labs(x = "FST", y = "", color = "category") + theme(legend.position = "top")
```

Correlation between Fst and distance between areas overall

```{r, fig.height=5, fig.width=5}
xx %>% mutate(Function2 = if_else(Function == "Rest", "Rest", "Mitochondrial")) %>% group_by(label2, Function2) %>% summarise(Fst = mean(Fst), distance = mean(distance)) %>% filter(!label2 %in% c("West vs West", "East vs East")) %>% ggscatter(x = "Fst", y = "distance",color = "Function2", add = "reg.line") + stat_cor(method = "pearson", label.x.npc =  "left", label.y.npc = "top", size = 2) + theme(legend.position = "", strip.text = element_text(size = 7))  + geom_text_repel(aes(label = label2), size = 2)
```

Correlation between Fst and distance between areas for each category

```{r, fig.height=10, fig.width=15}
pdf("../Figures/mtcategories_clinesv2.pdf", width = 10, height = 6)
mtclines = xx %>% mutate(Function2 = if_else(Function == "Rest", "Non-MT", "Mitochondrial")) %>% group_by(label2, Function) %>% summarise(Fst = mean(Fst), distance = mean(distance)) %>% filter(!label2 %in% c("West vs West", "East vs East"))
mtclines$Function = factor(mtclines$Function, levels = c(unique(mtclines$Function)[!grepl("Rest", unique(mtclines$Function))], "Rest"))

mtclines = mtclines %>% data.frame() %>% mutate(color = if_else(Function %in% c("ERMES complex", "fatty acid beta-oxidation", "SLC25 Family", "mitochondrial transcription and translation", "mitochondrial transport"), "West less divergent from Selfe than SC", "West more divergent from Selfe than SC"))
line_labels <- mtclines %>% data.frame() %>%
    select(distance, label2) %>%
    distinct()

line_segments <- mtclines %>%
    arrange(Function, distance) %>%
    group_by(Function, color) %>%  # Include the facetting variable!
    mutate(
        next_distance = lead(distance),
        next_Fst = lead(Fst)
    ) %>%
    filter(!is.na(next_distance), (next_distance - distance) <= 50) %>%
    ungroup()

ggplot(mtclines, aes(x = Fst, y = distance)) +
    # Draw only short lines between points
    geom_segment(data = line_segments,
                 aes(x = Fst, xend = next_Fst,
                     y = distance, yend = next_distance,
                     color = Function),
                 linewidth = 0.8,
                 inherit.aes = FALSE) +
    geom_point(aes(color = Function), size = 2) +
    facet_grid(. ~ color) +
    geom_hline(yintercept = unique(mtclines$distance), 
               linetype = "dashed", color = "black", alpha = 0.5) +
    geom_text(data = line_labels,
              aes(x = 0, y = distance, label = label2),
              vjust = -0.5, hjust = 0, size = 3, inherit.aes = FALSE) +
    theme_bw() +
    theme(
        strip.text = element_text(face = "bold"),
        legend.position = "right"
    ) +
    labs(
        y = "Geographic Distance (km)",
        x = expression(F[ST])
    ) + scale_fill_manual(values = funcols) + scale_color_manual(values = funcols) + guides(fill=guide_legend(ncol=1), color=guide_legend(ncol=1))
dev.off()
```

Correlation between Fst and distance between lakes for each category 

```{r, fig.height=12, fig.width=15}
xx %>% mutate(Function2 = if_else(Function == "Rest", "Rest", "Mitochondrial")) %>% group_by(label, Function) %>% summarise(Fst = mean(Fst), distance = mean(distance)) %>% filter(!label %in% c("Alexandrina_Middleton", "Mapourika_Paringa")) %>% ggscatter(x = "distance", y = "Fst",color = "Function", add = "reg.line") + stat_cor(method = "pearson", label.x.npc =  "left", label.y.npc = "top", size = 2) + theme(legend.position = "", strip.text = element_text(size = 7)) + scale_color_manual(values = brewPalette(n = 22)) + facet_wrap(~Function,ncol = 5) + geom_text_repel(aes(label = label), size = 2) 
```

Manhattan plot of the mitochondrial functions for East vs Selfe

```{r, fig.height=5, fig.width=15}
xxx = xxx %>% mutate(Function2_color = case_when(Function2 == "Mitochondrial" & color == "not_outlier"~ "Mitochondrial not_outlier", Function2 == "Mitochondrial" & color == "final_outlier" ~ Function, .default = "rest")) 
xxx$Function2_color = factor(xxx$Function2_color, levels = c("Mitochondrial not_outlier", unique(xxx$Function2_color)[c(-1,-2)], "rest"))
xxx$label = factor(xxx$label, levels = c("Alexandrina_Mapourika", "Alexandrina_Paringa", "Mapourika_Middleton", "Middleton_Paringa", "Alexandrina_Selfe", "Middleton_Selfe", "Paringa_Selfe", "Mapourika_Selfe", "Alexandrina_Middleton","Mapourika_Paringa"))
xxx = xxx[xxx$label %in% c("Alexandrina_Selfe", "Middleton_Selfe", "Paringa_Selfe", "Mapourika_Selfe"),]

Cb64k <- c(funcols, setNames(c("grey", "black"), c("rest", "Mitochondrial not_outlier")))

pdf("../Figures/Manhattan_Mitochondrial.pdf",height = 8, width = 10)
ggplot(xxx,aes(window,Fst)) + geom_point(data = xxx[xxx$Function2_color == "rest",], aes(window,Fst), color = "grey") + geom_point(data = xxx[xxx$Function2_color == "Mitochondrial not_outlier",], aes(window,Fst), color = "black") + geom_point(data = xxx[!(xxx$Function2_color %in% c("Mitochondrial not_outlier", "rest")),], aes(window,Fst, color = Function2_color), size = 3) + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) + facet_wrap(~label, ncol = 1) + scale_color_manual(values = funcols) + labs(color = "N-MT gene category", y = "FST", x = "window") + theme(axis.title.y = element_text(size = 15), strip.text = element_text(size = 15), axis.title.x =  element_text(size = 15))
dev.off()
```

Venn - how many overlap with bottom up analysis

```{r, fig.height=10, fig.width=15}

list_of_outliers = NULL
for (i in unique(for_plots$label)){
  tb = for_plots[for_plots$label == i & for_plots$color == "final_outlier",]
  tb = tb$window
  list_of_outliers[[i]] = tb
}

West_East_outliers = setdiff(setdiff(intersect(intersect(intersect(list_of_outliers$Alexandrina_Paringa, list_of_outliers$Alexandrina_Mapourika), list_of_outliers$Mapourika_Middleton), list_of_outliers$Middleton_Paringa), list_of_outliers$Alexandrina_Middleton), list_of_outliers$Mapourika_Paringa)

Incoming_to_Selfe_East= setdiff(setdiff(intersect(intersect(West_East_outliers, list_of_outliers$Mapourika_Selfe), list_of_outliers$Paringa_Selfe), list_of_outliers$Alexandrina_Selfe), list_of_outliers$Middleton_Selfe)
Incoming_to_Selfe_East=Incoming_to_Selfe_East[!duplicated(Incoming_to_Selfe_East)]
  
Incoming_to_Selfe_West= setdiff(setdiff(intersect(intersect(West_East_outliers, list_of_outliers$Middleton_Selfe), list_of_outliers$Alexandrina_Selfe), list_of_outliers$Mapourika_Selfe),list_of_outliers$Paringa_Selfe)
Incoming_to_Selfe_West=Incoming_to_Selfe_West[!duplicated(Incoming_to_Selfe_West)]

BDMI = setdiff(setdiff(setdiff(setdiff(West_East_outliers, list_of_outliers$Alexandrina_Selfe), list_of_outliers$Mapourika_Selfe), list_of_outliers$Paringa_Selfe), list_of_outliers$Middleton_Selfe)

obj = for_plots %>% 
  mutate(category = case_when(window %in% Incoming_to_Selfe_East ~ "Introgressing from South Canterbury", 
                           window %in% Incoming_to_Selfe_West ~ "Introgressing from West",
                           window %in% BDMI ~ "Introgressing from both",
                           .default = "Rest")) 
Incoming_to_Selfe_East_genes = setdiff(unique(obj[obj$category == "Introgressing from South Canterbury",]$gene_name), "")
Incoming_to_Selfe_West_genes = setdiff(unique(obj[obj$category == "Introgressing from West",]$gene_name), "")
BDMI_genes = setdiff(unique(obj[obj$category == "Introgressing from both",]$gene_name), "")

Mitochondrial=unique(list_of_mt_categories$Gene)

pdf("../Figures/GGvenn_mito_WestvsEast_overlap.pdf", height = 6, width = 9)
ggVennDiagram(list("Introgressing from both" = BDMI_genes, "Introgressing from South Canterbury"= Incoming_to_Selfe_East_genes, "Introgressing from West" = Incoming_to_Selfe_West_genes, "Mitochondrial" = Mitochondrial), set_color = c("orange", "plum", "darkcyan", "black")) + scale_fill_gradient2(midpoint = 150, low = "skyblue4", high = "pink4")  + scale_x_continuous(expand = expansion(mult = .4))
dev.off()
```

```{r, fig.height=10, fig.width=15}

pdf("../Figures/GGvenn_mito_overlap.pdf", height = 3, width = 5)
cats =list("South Canterbury vs Selfe" = unique(xx[(xx$label2 == "South Canterbury vs Selfe") & (xx$color == "final_outlier") & (xx$Function2 == "Mitochondrial"),]$window), "West vs Selfe" = unique(xx[(xx$label2 == "West vs Selfe") & (xx$color == "final_outlier") & (xx$Function2 == "Mitochondrial"),]$window), "West vs South Canterbury" = unique(xx[(xx$label2 == "West vs South Canterbury") & (xx$color == "final_outlier") & (xx$Function2 == "Mitochondrial"),]$window))
ggVennDiagram(cats) + scale_x_continuous(expand = expansion(mult = .4)) + ggtitle("Diverget genomic windows within N-MT genes") 
dev.off()
```

Overlap in the outliers between Alexandrina Selfe and Middleton Selfe comparisons

Windows

```{r, fig.height=10, fig.width=15}

ggVennDiagram(list("Alex_Selfe" = xxx[xxx$label == "Alexandrina_Selfe" & xxx$color == "final_outlier" & xxx$Function2 == "Mitochondrial",]$window, "Middleton_Selfe" = xxx[xxx$label == "Middleton_Selfe" & xxx$color == "final_outlier" & xxx$Function2 == "Mitochondrial",]$window))

```

Genes

```{r, fig.height=10, fig.width=15}

ggVennDiagram(list("Alex_Selfe" = xxx[xxx$label == "Alexandrina_Selfe" & xxx$color == "final_outlier" & xxx$Function2 == "Mitochondrial",]$gene_name, "Middleton_Selfe" = xxx[xxx$label == "Middleton_Selfe" & xxx$color == "final_outlier" & xxx$Function2 == "Mitochondrial",]$gene_name))

```













